### PR TITLE
Init logic to retry http request if there are connection problem

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
Hi,

In this PR I propose the idea described inside this issue #16. In most of cases the connection can have problems, e.g: The VPN connection crashed during a esplora fetch operation, and this cause the crash of lightningd. Or the esplora don't have a feerate available for the last block and with this code, the plugin can wait before to retry the operation.

In this PR I wrote the not maintainable code but I promise that if this logic work I will refactoring it with a call back function :-)

I hope this can help!

Signed-off-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

### CODE REVIEW

File: esplora.c [hunk 1]
```diff
@@ -76,47 +79,78 @@ static size_t write_memory_callback(void *contents, size_t size, size_t nmemb,
...
                    const char *data) {
+
   struct curl_memory_data chunk;
+  const struct command *cmd = (struct command*) ctx;
+  u64 time = 0;
+  
   chunk.memory = tal_arr(ctx, u8, 64);
   chunk.size = 0;
-
+
```
Comment (darosior) [inline esplora.c:89]: Trailing space

File: esplora.c [hunk 2]
```diff
@@ -76,47 +79,78 @@ static size_t write_memory_callback(void *contents, size_t size, size_t nmemb,
 
 static u8 *request(const tal_t *ctx, const char *url, const bool post,
                    const char *data) {
+
```
Comment (darosior) [inline esplora.c:82]: spurious CR insertion ?

File: esplora.c [hunk 3]
```diff
@@ -76,47 +79,78 @@ static size_t write_memory_callback(void *contents, size_t size, size_t nmemb,
...
+  
   chunk.memory = tal_arr(ctx, u8, 64);
   chunk.size = 0;
-
+  
   CURL *curl;
   CURLcode res;
   curl = curl_easy_init();
   if (!curl) {
+    plugin_log(cmd->plugin, LOG_UNUSUAL, "Error with lincurl init call");
```
Comment (darosior) [inline esplora.c:94]:
```suggestion
    plugin_log(cmd->plugin, LOG_BROKEN, "Could not get a curl context");
```

File: esplora.c [hunk 4]
```diff
@@ -76,47 +79,78 @@ static size_t write_memory_callback(void *contents, size_t size, size_t nmemb,
 
 static u8 *request(const tal_t *ctx, const char *url, const bool post,
                    const char *data) {
+
   struct curl_memory_data chunk;
+  const struct command *cmd = (struct command*) ctx;
```
Comment (darosior) [inline esplora.c:84]: Ouch. No, a `tal_t` struct cannot be cast to a `command` one!

File: esplora.c [hunk 5]
```diff
@@ -76,47 +79,78 @@ static size_t write_memory_callback(void *contents, size_t size, size_t nmemb,
...
   res = curl_easy_perform(curl);
+
   if (res != CURLE_OK) {
+    if (time < retry_time){
+      time++;
+      plugin_log(cmd->plugin, LOG_DBG, "Curl return a result different to CURLE_OK, it is %d", CURLE_OK);
+      sleep(wait_time);
+      plugin_log(cmd->plugin, LOG_DBG, "%ld: Retry curl request to URL: %s", time, url);
+      goto retrycurl;
+    }
```
Comment (darosior) [inline esplora.c:128]: Hmm. I propose:
```C
for (;;) {
    res = curl_easy_perform(curl);
    if (res == CURLE_OK)
        break;
    if (++time > retry_time)
        return tal_free(chunk.memory);
    sleep(wait_time);
}
```

File: esplora.c [hunk 6]
```diff
@@ -76,47 +79,78 @@ static size_t write_memory_callback(void *contents, size_t size, size_t nmemb,
...
+  retryinfo:
   curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
   if (response_code != 200) {
+    if(time < retry_time){
+      time++;
+      plugin_log(cmd->plugin, LOG_DBG, "bad code in the info answare! wait %d before retry", wait_time);
+      sleep(wait_time);
+      plugin_log(cmd->plugin, LOG_DBG, "%ld: Retry info request to URL=%s", time, url);
+      goto retryinfo;
+    }
```
Comment (darosior) [inline esplora.c:145]: Just make my above proposition a function ?

File: esplora.c [hunk 7]
```diff
@@ -20,11 +20,14 @@
 #include <errno.h>
 #include <inttypes.h>
 #include <plugins/libplugin.h>
+#include <unistd.h>
 
 static char *endpoint = NULL;
 static char *cainfo_path = NULL;
 static char *capath = NULL;
 static u64 verbose = 0;
+static int wait_time = 60; //Time express in second
```
Comment (darosior) [inline esplora.c:29]: This is *way* too long. I think it can just be a hardcoded value in a `make_curl_request` function ?

File: esplora.c [hunk 8]
```diff
@@ -76,47 +79,78 @@ static size_t write_memory_callback(void *contents, size_t size, size_t nmemb,
...
+    }
     return tal_free(chunk.memory);
   }
   curl_easy_cleanup(curl);
   tal_resize(&chunk.memory, chunk.size);
   return chunk.memory;
 }
 
-static char *request_get(const tal_t *ctx, const char *url) {
+ static char *request_get(const tal_t *ctx, const char *url) {
```
Comment (darosior) [inline esplora.c:153]: spurious space ?

File: esplora.c [hunk 9]
```diff
@@ -148,7 +182,7 @@ static struct command_result *getchaininfo(struct command *cmd,
                                            const char *buf UNUSED,
                                            const jsmntok_t *toks UNUSED) {
   char *err;
-
+
```
Comment (darosior) [inline esplora.c:185]: spurious space ?

**Review by darosior — CHANGES_REQUESTED** [review]

**Comment by darosior** [comment]
Do you mind if i propose an alternate PR ?

**Comment by vincenzopalazzo** [comment]
>Do you mind if i propose an alternate PR ?

@darosior Absolutly no! I know that my C is very bad at the moment! I can not wait to see you PR :-)

File: esplora.c (hunk 7 above)
Comment (vincenzopalazzo) [reply esplora.c:29]: Do you mean the time is too long?

File: esplora.c (hunk 5 above)
Comment (vincenzopalazzo) [reply esplora.c:128]: Very cool :-) I was too excited to use the goto :-P

File: esplora.c (hunk 4 above)
Comment (vincenzopalazzo) [reply esplora.c:84]: Ops! Is i want the log here I could add another parameter to the request? only solution?

**Pull request closed without merging**